include/rmm/mr/device/system_memory_resource.hpp

```diff
@@ -0,0 +1,187 @@
+    if (headroom_size_ == 0) {
+      // If headroom is not specified, always pass through to malloc.
+      threshold_size_ = threshold_size.value_or(std::numeric_limits<std::size_t>::max());
+    } else {
+      // If headroom is specified, by default we check every allocation.
+      threshold_size_ = threshold_size.value_or(0);
+    }
+  }
+
+  ~system_memory_resource() override                    = default;
+  system_memory_resource(system_memory_resource const&) = default;  ///< @default_copy_constructor
+  system_memory_resource(system_memory_resource&&)      = default;  ///< @default_move_constructor
+  system_memory_resource& operator=(system_memory_resource const&) =
+    default;  ///< @default_copy_assignment{system_memory_resource}
+  system_memory_resource& operator=(system_memory_resource&&) =
+    default;  ///< @default_move_assignment{system_memory_resource}
+
+ private:
+  /**
+   * @brief Allocates memory of size at least \p bytes.
+   *
+   * The returned pointer will have at minimum 256 byte alignment.
+   *
+   * The stream argument is ignored.
+   *
+   * @param bytes The size of the allocation
+   * @param stream This argument is ignored
+   * @return void* Pointer to the newly allocated memory
+   */
+  void* do_allocate(std::size_t bytes, [[maybe_unused]] cuda_stream_view stream) override
+  {
+    auto const aligned{rmm::align_up(bytes, rmm::CUDA_ALLOCATION_ALIGNMENT)};
+    void* ptr{malloc(aligned)};
+
+    if (bytes >= threshold_size_) {
+      auto const free        = rmm::available_device_memory().first;
+      auto const allocatable = std::max(free - headroom_size_, 0UL);
+      auto const gpu_portion = std::min(allocatable, aligned);
+      auto const cpu_portion = aligned - gpu_portion;
+      if (gpu_portion != 0) {
+        RMM_CUDA_TRY(cudaMemAdvise(ptr,
+                                   gpu_portion,
+                                   cudaMemAdviseSetPreferredLocation,
+                                   rmm::get_current_cuda_device().value()));
+      }
+      if (cpu_portion != 0) {
+        RMM_CUDA_TRY(cudaMemAdvise(static_cast<char*>(ptr) + gpu_portion,
+                                   cpu_portion,
+                                   cudaMemAdviseSetPreferredLocation,
+                                   cudaCpuDeviceId));
+      }
+    }
+    return ptr;
+  }
+
+  /**
+   * @brief Deallocate memory pointed to by \p p.
+   *
+   * The stream argument is ignored.
+   *
+   * @param ptr Pointer to be deallocated
+   * @param bytes This argument is ignored
+   * @param stream This argument is ignored
+   */
+  void do_deallocate(void* ptr,
+                     [[maybe_unused]] std::size_t bytes,
+                     [[maybe_unused]] cuda_stream_view stream) override
+  {
+    free(ptr);
+  }
+
+  /**
+   * @brief Compare this resource to another.
+   *
+   * Two cuda_memory_resources always compare equal, because they can each
+   * deallocate memory allocated by the other.
```

Comment:
*issue*: This isn't a `cuda_memory_resource` though.

I _think_ for equality, the requirement RMM tends to have is that pointers allocated by memory resource A can be deallocated by memory resource B. But here we are stricter, since we also require that the headroom and threshold are identical. Is this desired?

Reply:
Fixed.